File: tools/spore/rw4/rw4.py
#!/usr/bin/env python3
"""RW4 container parser (Spore variant) — header, manifest, type codes, section infos, object type codes."""
import struct
import sys

MAGIC = bytes([0x89, 0x52, 0x57, 0x34, 0x77, 0x33, 0x32, 0x00, 0x0D, 0x0A, 0x1A, 0x0A, 0x00, 0x20, 0x04, 0x00,
               0x34, 0x35, 0x34, 0x00, 0x30, 0x30, 0x30, 0x00, 0x00, 0x00, 0x00, 0x00])


def u32(b, o):
    return struct.unpack_from('<I', b, o)[0]


def i32(b, o):
    return struct.unpack_from('<i', b, o)[0]


TYPES = {0x10030: 'BaseResource', 0x10031: '?', 0x10032: '?', 0x10010: '?', 0x10004: 'SectionManifest',
         0x10005: 'SectionTypes', 0x10006: 'SectionExternalArenas', 0x10007: 'SectionSubReferences',
         0x10008: 'SectionAtoms', 0x20003: 'Raster', 0x20001: 'VertexDescription', 0x20002: 'VertexBuffer',
         0x20004: 'IndexBuffer', 0x20005: 'SkinMatrixBuffer', 0x30001: 'AnimationSkin', 0x40001: 'Mesh',
         0x50001: 'MeshCompiledStateLink', 0x60001: 'CompiledState', 0x70001: 'SkinsInK',
         0x80001: 'SkeletonsInK', 0x80002: 'Skeleton', 0x80003: 'BBox', 0x90001: 'MorphHandle',
         0x90002: 'TriangleKDTreeProcedural', 0xA0001: 'Animations', 0xA0002: 'KeyframeAnim',
         0xB0001: 'BlendShape', 0xB0002: 'BlendShapeBuffer', 0xC0001: 'TextureOverride'}


def parse(path):
    with open(path, 'rb') as f:
        b = f.read()
    assert b[:28] == MAGIC, f"bad magic {b[:12].hex()}"
    ftype = u32(b, 0x1C)
    tname = {1: 'MODEL', 0x04000000: 'TEXTURE', 0xCAFED00D: 'SPECIAL'}.get(ftype, f"0x{ftype:08x}")
    objCount = u32(b, 0x20); sectionCount = u32(b, 0x24)
    f28 = u32(b, 0x28); f2C = u32(b, 0x2C)
    pSectionInfo = u32(b, 0x30); f34 = u32(b, 0x34)
    pBufferData = u32(b, 0x44); f48 = u32(b, 0x48); bufSize = u32(b, 0x4C)
    print(f"### {path}")
    print(f"type={tname} objCount={objCount} sectionCount={sectionCount} f28={f28} pSectionInfo=0x{pSectionInfo:x} pBufferData=0x{pBufferData:x} bufSize={bufSize}")
    # manifest at 0x98
    mo = 0x98
    assert u32(b, mo) == 0x10004, f"manifest tc 0x{u32(b, mo):x}"
    f0, f4 = i32(b, mo + 4), i32(b, mo + 8)
    o1, o2, o3, o4 = u32(b, mo + 0xC), u32(b, mo + 0x10), u32(b, mo + 0x14), u32(b, mo + 0x18)
    # types
    to = mo + o1
    assert u32(b, to) == 0x10005, f"types tc 0x{u32(b, to):x}"
    n = u32(b, to + 4)
    codes = [u32(b, to + 8 + 4 * i) for i in range(n)]
    print(f"manifest: f0={f0} f4={f4} types@0x{to:x}(n={n}) arenas@0x{mo + o2:x} subref@0x{mo + o3:x} atoms@0x{mo + o4:x}")
    print("typecodes: " + ", ".join(f"0x{c:x}:{TYPES.get(c, '?')}" for c in codes))
    # section infos
    print("sections:")
    for i in range(sectionCount):
        o = pSectionInfo + 24 * i
        pData, f04, size, align, tci, tc = u32(b, o), u32(b, o + 4), i32(b, o + 8), i32(b, o + 0xC), i32(b, o + 0x10), i32(b, o + 0x14)
        ap = pData + pBufferData if tc == 0x10030 else pData
        print(f"  [{i}] tc=0x{tc:x}:{TYPES.get(tc, '?')} tcIdx={tci} size={size} align={align} data@0x{ap:x}")
    return b


if __name__ == '__main__':
    for p in sys.argv[1:]:
        parse(p)
        print()

File: src/assets/Rw4.cpp
// See Rw4.hpp. Clean-room implementation, independently authored.
#include "Rw4.hpp"

#include <cstdio>
#include <limits>
#include <utility>

#include "Stream.hpp"

namespace openspore::assets {

namespace {

constexpr uint8_t kMagic[kRw4MagicSize] = {
    0x89, 0x52, 0x57, 0x34, 0x77, 0x33, 0x32, 0x00, 0x0D, 0x0A,
    0x1A, 0x0A, 0x00, 0x20, 0x04, 0x00, 0x34, 0x35, 0x34, 0x00,
    0x30, 0x30, 0x30, 0x00, 0x00, 0x00, 0x00, 0x00,
};

// Documented RW4 object / section type codes (SMFX map, docs/
// RENDERWARE-RESEARCH.md §7.5). Used only to count *undocumented* codes (e.g.
// Spore's 0x7000c / 0x7000f); it is NOT a structural gate.
bool inKnownSet(uint32_t tc) {
  switch (tc) {
    case 0x10030:   // BaseResource
    case 0x10004:   // SectionManifest
    case 0x10005:   // SectionTypes
    case 0x10006:   // SectionExternalArenas
    case 0x10007:   // SectionSubReferences
    case 0x10008:   // SectionAtoms
    case 0x20003:   // Raster
    case 0x20007:   // IndexBuffer
    case 0x20008:   // TextureOverride
    case 0x20009:   // Mesh
    case 0x2001a:   // MeshCompiledStateLink
    case 0x200af:   // BlendShapeBuffer
    case 0x2000b:   // (Spore-specific, unmapped)
    case 0x70001:   // KeyframeAnim
    case 0x70002:   // Skeleton
    case 0x70003:   // AnimationSkin
    case 0x7000b:   // (Spore-specific, unmapped)
    case 0x80003:   // TriangleKDTreeProcedural
    case 0x80005:   // BBox
    case 0xff0000:  // MorphHandle
    case 0xff0001:  // Animations
    case 0xff0002:  // BlendShape
      return true;
    default:
      return false;
  }
}

// u32 -> "0x<hex>" (lowercase, no leading zeros) for the canonical summary.
std::string hexNoPad(uint32_t v) {
  char buf[16];
  std::snprintf(buf, sizeof(buf), "0x%x", v);
  return buf;
}

std::string hexPlain(uint32_t v) {
  char buf[16];
  std::snprintf(buf, sizeof(buf), "%x", v);
  return buf;
}

}  // namespace

bool knownRw4TypeCode(uint32_t tc) { return inKnownSet(tc); }

bool Rw4::complete() const {
  if (sections.size() != sectionCount) {
    return false;
  }
  for (const Rw4Section& s : sections) {
    if (s.size < 0) {
      return false;
    }
    const uint64_t end =
        static_cast<uint64_t>(s.data) + static_cast<uint64_t>(s.size);
    if (end > recordSize) {
      return false;
    }
  }
  return true;
}

size_t Rw4::unknownTypeCodes() const {
  size_t n = 0;
  for (const Rw4Section& s : sections) {
    if (!inKnownSet(s.tc)) {
      ++n;
    }
  }
  return n;
}

std::string Rw4::describe() const {
  std::string line = hexNoPad(ftype) + " obj=" + std::to_string(objCount) +
                     " sec=" + std::to_string(sectionCount) +
                     " buf=" + std::to_string(bufSize);
  for (const Rw4Section& s : sections) {
    line += " | " + hexPlain(s.tc) + " d=" + hexNoPad(s.data) +
            " s=" + std::to_string(s.size);
  }
  return line;
}

bool parseRw4(const uint8_t* data, size_t size, Rw4& out, std::string& error) {
  out = Rw4{};
  error.clear();
  Rw4 parsed;
  parsed.recordSize = size;

  if (data == nullptr) {
    error = "rw4: null input";
    return false;
  }
  if (size < kRw4MagicSize + 0x78) {
    error =
        "record too small for RW4 header (" + std::to_string(size) + " bytes)";
    return false;
  }
  for (size_t i = 0; i < kRw4MagicSize; ++i) {
    if (data[i] != kMagic[i]) {
      error = "bad RW4 magic";
      return false;
    }
  }

  // Header words (absolute offsets).
  parsed.ftype =
      static_cast<uint32_t>((static_cast<uint32_t>(data[0x1C]) |
                             (static_cast<uint32_t>(data[0x1D]) << 8) |
                             (static_cast<uint32_t>(data[0x1E]) << 16) |
                             (static_cast<uint32_t>(data[0x1F]) << 24)));
  auto u32at = [&](size_t o) {
    return static_cast<uint32_t>(data[o]) |
           (static_cast<uint32_t>(data[o + 1]) << 8) |
           (static_cast<uint32_t>(data[o + 2]) << 16) |
           (static_cast<uint32_t>(data[o + 3]) << 24);
  };
  parsed.objCount = u32at(0x20);
  parsed.sectionCount = u32at(0x24);
  parsed.pSectionInfo = u32at(0x30);
  parsed.pBufferData = u32at(0x44);
  parsed.bufSize = u32at(0x4C);

  if (parsed.pBufferData > size || parsed.bufSize > size - parsed.pBufferData) {
    error = "buffer arena out of bounds";
    return false;
  }
  if (parsed.ftype != Rw4::kTypeModel && parsed.ftype != Rw4::kTypeTexture &&
      parsed.ftype != Rw4::kTypeSpecial) {
    error = "unknown ftype 0x" + hexPlain(parsed.ftype);
    return false;
  }

  // Manifest at 0x98.
  const size_t mo = 0x98;
  if (mo + 0x1C > size) {
    error = "record too small for manifest";
    return false;
  }
  const uint32_t manifestTc = u32at(mo);
  if (manifestTc != Rw4::kManifestTc) {
    error = "manifest typecode 0x" + hexPlain(manifestTc);
    return false;
  }
  const uint32_t o1 = u32at(mo + 0x0C);  // -> SectionTypes
  const uint64_t typesOffValue = static_cast<uint64_t>(mo) + o1;
  if (typesOffValue + 8u > size) {
    error = "SectionTypes out of bounds";
    return false;
  }
  const size_t typesOff = static_cast<size_t>(typesOffValue);
  const uint32_t typesTc = u32at(typesOff);
  if (typesTc != Rw4::kTypesTc) {
    error = "types typecode 0x" + hexPlain(typesTc);
    return false;
  }
  const uint32_t nTypes = u32at(typesOff + 4);
  const uint64_t typesEnd =
      typesOffValue + 8u + static_cast<uint64_t>(nTypes) * 4u;
  if (typesEnd > size) {
    error = "typecode table out of bounds (n=" + std::to_string(nTypes) + ")";
    return false;
  }
  parsed.typecodes.reserve(nTypes);
  for (uint32_t i = 0; i < nTypes; ++i) {
    parsed.typecodes.push_back(
        u32at(typesOff + 8u + static_cast<size_t>(i) * 4u));
  }

  // Section infos.
  if (parsed.pSectionInfo > size ||
      static_cast<uint64_t>(parsed.sectionCount) * 24u >
          size - parsed.pSectionInfo) {
    error = "section-info table out of bounds";
    return false;
  }
  parsed.sections.reserve(parsed.sectionCount);
  for (uint32_t i = 0; i < parsed.sectionCount; ++i) {
    const size_t o =
        static_cast<size_t>(parsed.pSectionInfo) + static_cast<size_t>(i) * 24u;
    Rw4Section s;
    s.pData = u32at(o);
    s.f04 = u32at(o + 4);
    s.size = static_cast<int32_t>(u32at(o + 8));
    s.align = static_cast<int32_t>(u32at(o + 0xC));
    s.tcIndex = static_cast<int32_t>(u32at(o + 0x10));
    s.tc = static_cast<uint32_t>(u32at(o + 0x14));
    const uint64_t data =
        s.tc == 0x10030 ? static_cast<uint64_t>(s.pData) + parsed.pBufferData
                        : s.pData;
    if (data > std::numeric_limits<uint32_t>::max()) {
      error = "section data address overflows";
      return false;
    }
    s.data = static_cast<uint32_t>(data);
    parsed.sections.push_back(s);
  }

  if (!parsed.complete()) {
    error = "section walk not well-formed (bad type code / out-of-bounds data)";
    return false;
  }
  out = std::move(parsed);
  return true;
}

}  // namespace openspore::assets
